File: src/ice/world/blocks/liquids/LiquidClassifier.java
package ice.world.blocks.liquids;

import arc.graphics.g2d.Draw;
import arc.graphics.g2d.TextureRegion;
import arc.scene.ui.layout.Table;
import arc.util.Eachable;
import arc.util.io.Reads;
import arc.util.io.Writes;
import mindustry.entities.units.BuildPlan;
import mindustry.gen.Building;
import mindustry.type.Liquid;
import mindustry.world.Block;
import mindustry.world.blocks.ItemSelection;
import mindustry.world.blocks.liquid.LiquidBlock;
import mindustry.world.meta.Stat;

import static ice.library.drawf.IceDraw.*;
import static mindustry.Vars.content;

public class LiquidClassifier extends Block {
    public TextureRegion top = LoadText(name + "-top");
    public TextureRegion bottom = LoadText(name + "-bottom");

    public LiquidClassifier(String name) {
        super(name);
        buildType = LiquidClassifierBuild::new;
        liquidCapacity = 10;
        hasLiquids = true;
        destructible = true;
        underBullets = true;
        update = true;
        instantTransfer = true;
        configurable = true;
        saveConfig = true;
        clearOnDoubleTap = true;
        outputsLiquid = true;
        config(Liquid.class, (LiquidClassifier.LiquidClassifierBuild tile, Liquid liquid)->tile.sortLiquid = liquid);
        configClear((LiquidClassifier.LiquidClassifierBuild tile)->tile.sortLiquid = null);
    }

    @Override
    public void setStats() {
        super.setStats();
        stats.remove(Stat.maxConsecutive);
    }

    @Override
    public void drawPlanConfig(BuildPlan plan, Eachable<BuildPlan> list) {
        drawPlanConfigCenter(plan, plan.config, name + "-top2");
    }

    @Override
    public void drawPlanConfigCenter(BuildPlan plan, Object content, String region) {
         if(content instanceof Liquid l){
             Draw.color(l.color);
             Draw.rect(region, plan.drawx(), plan.drawy());
             Draw.color();
             Draw.rect(top, plan.drawx(), plan.drawy());
         }
    }

    public class LiquidClassifierBuild extends Building {
        public Liquid sortLiquid;

        @Override
        public void draw() {
            Draw.rect(bottom, x, y);
            if (sortLiquid != null) {
                LiquidBlock.drawTiledFrames(size, x, y, 0f, sortLiquid, 1f);
            }
            Draw.rect(top, x, y);
        }

        /** 复制蓝图会调用 */
        @Override
        public Liquid config() {
            return sortLiquid;
        }

        @Override
        public void buildConfiguration(Table table) {
            ItemSelection.buildTable(block, table, content.liquids(), ()->sortLiquid, this::configure, selectionRows, selectionColumns);
            super.buildConfiguration(table);
        }

        @Override
        public boolean acceptLiquid(Building source, Liquid liquid) {
            return liquid == sortLiquid;
        }

        @Override
        public void updateTile() {
            if (sortLiquid != null) {
                dumpLiquid(sortLiquid);
            }
        }

        @Override
        public void write(Writes write) {
            super.write(write);
            write.s(sortLiquid == null ? -1 : sortLiquid.id);
        }

        @Override
        public void read(Reads read, byte revision) {
            super.read(read, revision);
            int id = revision == 1 ? read.s() : read.b();
            sortLiquid = id == -1 ? null : content.liquid(id);
        }
    }
}

File: src/ice/world/content/blocks/liquid/LiquidClassifier.kt
package ice.world.content.blocks.liquid

import arc.func.Prov
import arc.graphics.g2d.TextureRegion
import arc.scene.ui.layout.Table
import arc.util.Eachable
import arc.util.io.Reads
import arc.util.io.Writes
import ice.library.scene.ui.ItemSelection
import ice.library.struct.texture.LazyTextureSingleDelegate
import ice.world.draw.DrawBuild
import ice.world.draw.DrawMulti
import mindustry.Vars
import mindustry.entities.units.BuildPlan
import mindustry.gen.Building
import mindustry.graphics.Drawf
import mindustry.type.Liquid
import mindustry.world.blocks.liquid.LiquidBlock
import mindustry.world.draw.DrawRegion
import mindustry.world.meta.BlockGroup
import mindustry.world.meta.Stat
import singularity.world.blocks.SglBlock

open class LiquidClassifier(name: String) : SglBlock(name) {
  val top2: TextureRegion by LazyTextureSingleDelegate("${this.name}-top2")

  init {
    update = true
    saveConfig = true
    hasLiquids = true
    configurable = true
    underBullets = true
    group = BlockGroup.liquids
    outputsLiquid = true
    clearOnDoubleTap = true
    buildType = Prov(::LiquidClassifierBuild)
    drawers = DrawMulti(DrawRegion("-bottom"), DrawBuild<LiquidClassifierBuild> {
      if (sortLiquid != null) {
        LiquidBlock.drawTiledFrames(size, x, y, 0f, sortLiquid, 1f)
      }
    }, DrawRegion("-top"))
    config(Liquid::class.java) { tile: LiquidClassifierBuild, liquid: Liquid? ->
      tile.sortLiquid = liquid
    }
    configClear { tile: LiquidClassifierBuild -> tile.sortLiquid = null }
  }

  override fun drawPlanConfig(plan: BuildPlan, list: Eachable<BuildPlan>) {
    val config = plan.config
    if (config is Liquid) {
      Drawf.liquid(top2, plan.drawx(), plan.drawy(), 0.6f, config.color)
    }
  }

  override fun setStats() {
    super.setStats()
    stats.remove(Stat.liquidCapacity)
  }

  inner class LiquidClassifierBuild : SglBuilding() {

    var sortLiquid: Liquid? = null
    override fun config(): Liquid? {
      return sortLiquid
    }

    override fun acceptLiquid(source: Building, liquid: Liquid)=false

    override fun buildConfiguration(table: Table) {
      ItemSelection.buildTable(block, table, Vars.content.liquids(), ::sortLiquid, this::configure, true)
    }

    override fun updateTile() {
      sortLiquid ?: return
      proximity.forEach {
        if (it is MultipleLiquidBlock.MultipleBlockBuild) {
          proximity.forEach { building ->

            it.moveLiquid(building, sortLiquid)

            /*if (building.acceptLiquid(this, sortLiquid)) {
              val amount: Float = min(it.liquids.get(sortLiquid), building.block.liquidCapacity - building.liquids.get(sortLiquid))
              building.handleLiquid(this, sortLiquid, amount)
              it.liquids.remove(sortLiquid, amount)
            }*/
          }
        }
      }
    }

    override fun read(read: Reads, revision: Byte) {
      super.read(read, revision)
      val id: Int = read.i()
      sortLiquid = if (id == -1) null else Vars.content.liquid(id)
    }

    override fun write(write: Writes) {
      super.write(write)
      write.i((sortLiquid?.id ?: -1).toInt())
    }
  }
}
